lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,376 @@
+            })
+
+            return cpus.issubset(fitting_cpus)
+
+        res = self._parse_phases(parse_phase, metrics)
+
+        self._plot_phases('test_placement', test_failures)
+
+        return res
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_freq_selection(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement and frequency selection
+        is compatible with UtilClamp requirements. This is done by comparing
+        the current CPU capacity on which the task has been placed, with the
+        UtilClamp value.
+
+        The expected capacity is the schedutil projected frequency selection
+        for the given uclamp value.
+        """
+
+        metrics = {}
+        test_failures = []
+        capacity_dfs = []
+        """
+        (
+            # schedutil factor that converts util to a frequency for a
+            # given CPU:
+            #
+            #   next_freq = max_freq * C * util / max_cap
+            #
+            #   where C = 1.25
+            schedutil_factor,
+
+            # list of frequencies available for a given CPU.
+            frequencies,
+        )
+        """
+        cpu_frequencies = {
+            cpu: (
+                (max(capacities) * (1 / self.CAPACITY_MARGIN)) /
+                max(capacities.values()),
+                sorted(capacities)
+            )
+            for cpu, capacities in
+            self._collect_capacities(self.plat_info).items()
+        }
+        cpu_capacities = self._collect_capacities(self.plat_info)
+
+        def schedutil_map_util_cap(cpu, util):
+            """
+            Returns, for a given util on a given CPU, the capacity that
+            schedutil would select.
+            """
+
+            schedutil_factor, frequencies = cpu_frequencies[cpu]
+            schedutil_freq = schedutil_factor * util
+
+            # Find the first available freq that meet the schedutil freq
+            # requirement.
+            for freq in frequencies:
+                if freq >= schedutil_freq:
+                    break
+
+            return cpu_capacities[cpu][freq]
+
+        def parse_phase(df, phase):
+            uclamp_val = phase['uclamp_val']
+            num_activations = df['active'][df['active'] == 1].count()
+            expected = schedutil_map_util_cap(df['cpu'].unique()[0],
+                                              uclamp_val)
+
+            # Activations numbering
+            df['activation'] = df['active'].cumsum()
+
+            # Only keep the activations
+            df.ffill(inplace=True)
+            df = df[df['active'] == 1]
+
+            # Actual capacity at which the task is running
+            for cpu, freq_to_capa in cpu_capacities.items():
+                df[cpu] = df[cpu].map(freq_to_capa)
+                df['capacity'] = df.apply(lambda line: line[line.cpu], axis=1)
```

Comment:
There seems to be something fishy here:
* we keep overwriting `df['capacity']` inside the loop
* we then use it after the loop is finished

Just taking it out of the loop should be enough, since it only relies on the columns labeled by CPU IDs, which are updated independently at each loop iteration.

Reply:
done.